### PR TITLE
JWT token flow

### LINKED ISSUE #242: JWT token flow
- If unauthorized, redirect to `/unauthorized` endpoint (FE)
- Store JWT tokens rather than sending thru header
- Fix tests
- Revert changes to protected endpoints introduced in #204

### PR DESCRIPTION
**Pull request by s278gupt**
Closes: https://github.com/uwblueprint/shoe-project/issues/242

By commit:

- Sets JWT token in httponly cookie `jwt`
- Redirects to frontend page `/unauthorized` if any of the following is true:
  - user tries to access protected endpoint without logging in
  - user email is not bp or tsp
  - jwt token is invalid/expired
- Reverts tests for protected endpoints

### CODE REVIEW

File: restapi/api.go [hunk 1]
```diff
@@ -62,32 +62,22 @@ func Router(db *gorm.DB, locationFinder location.LocationFinder) (http.Handler,
...
-		rest.PostHandler(r, "/story", api.CreateStoriesFormData)
-
 		rest.GetHandler(r, "/auth/callback", api.AuthCallback)
 		rest.GetHandler(r, "/client_tokens", api.ReturnClientTokens)
-		rest.GetHandler(r, "/unauthorized", api.UnauthorizedPage)
-
-		// TODO: move these back to private api once https://github.com/uwblueprint/shoe-project/issues/204 is closed
-		rest.PostHandler(r, "/stories", api.CreateStories)
-		rest.PostHandler(r, "/authors", api.CreateAuthors)
+		rest.PostHandler(r, "/story", api.CreateStoriesFormData)
```
Comment (AbhiPrasad) [inline restapi/api.go:70]: we probably want this to be protected as well

**Review by AbhiPrasad — APPROVED** [review]

File: restapi/api.go (hunk 1 above)
Comment (s278gupt) [reply restapi/api.go:70]: woops, good catch. ty

File: restapi/api.go [hunk 2]
```diff
@@ -62,32 +62,22 @@ func Router(db *gorm.DB, locationFinder location.LocationFinder) (http.Handler,
...
 
 	// Private API
 	r.Group(func(r chi.Router) {
 		r.Use(jwtauth.Verifier(config.GetJWTKey()))
-		r.Use(jwtauth.Authenticator)
+		r.Use(Authenticator)
 
-		// dummy call to check auth flow
-		rest.GetHandler(r, "/checkauth", api.checkAuthDummy)
+		rest.PostHandler(r, "/stories", api.CreateStories)
```
Comment (nimwijetunga) [inline restapi/api.go]: Move all of these to public for now so its easier to test (we can move them back in a different PR)

File: restapi/auth.go [hunk 3]
```diff
@@ -72,37 +73,62 @@ func (api api) AuthCallback(w http.ResponseWriter, r *http.Request) render.Rende
...
 }
 
-func generateJWTToken(email string) (string, error) {
+func generateJWTToken(email string, w http.ResponseWriter) error {
 	jwtExpiry, err := config.GetTokenExpiryDuration()
 	if err != nil {
-		return "", err
+		return err
 	}
+	expiration := time.Now().Add(jwtExpiry)
```
Comment (nimwijetunga) [inline restapi/auth.go:93]: whats the default expiry here?

File: restapi/auth.go [hunk 4]
```diff
@@ -72,37 +73,62 @@ func (api api) AuthCallback(w http.ResponseWriter, r *http.Request) render.Rende
...
 
 	token := config.GetJWTKey()
 	_, signedToken, err := token.Encode(claim)
 	if err != nil {
-		return "", err
+		return err
 	}
 
-	return signedToken, nil
+	secure := (config.GetMode() == config.MODE_PROD)
```
Comment (nimwijetunga) [inline restapi/auth.go:109]: what does this do?

File: restapi/endpoints_test.go [hunk 5]
```diff
@@ -30,9 +29,9 @@ func (suite *endpointTestSuite) SetupSuite() {
 	suite.db = db
 
 	// setup required for jwt authentication
-	viper.SetDefault("auth.jwt_key", "random")
-	viper.SetDefault("auth.jwt_expiry", "5m")
-	viper.SetDefault("auth.jwt_issuer", "endpoint_tests")
+	viper.SetDefault("auth.jwt_key", testutils.JWTKey)
```
Comment (nimwijetunga) [inline restapi/endpoints_test.go:32]: are these from the .env file in the regular app? If so can we add them to heroku as well

**Review by nimwijetunga — APPROVED** [review]
LGTM once you address the comments

File: restapi/auth.go (hunk 3 above)
Comment (s278gupt) [reply restapi/auth.go:93]: 2h

File: restapi/auth.go (hunk 4 above)
Comment (s278gupt) [reply restapi/auth.go:109]: if in prod, `Secure` field is set to true in cookie: this makes sure that cookie is only included in reqs over https

File: restapi/endpoints_test.go (hunk 5 above)
Comment (s278gupt) [reply restapi/endpoints_test.go:32]: no, these are from `testutils` file; the tests are written in isolation from rest of app, so we need to set env variables. i do this in the testutils file

File: restapi/auth.go (hunk 4 above)
Comment (nimwijetunga) [reply restapi/auth.go:109]: this may be an issue since we don't have https right now in prod (can address in a seperate pr if needed)

File: restapi/endpoints_test.go (hunk 5 above)
Comment (nimwijetunga) [reply restapi/endpoints_test.go:32]: okay as long as you have added any new env variables to heroku should be fine
